### DISCUSSION TITLE
using `queryRaw` with `sql` to do javascript functions?

### DISCUSSION THREAD

**Discussion by johhansantana** [prisma/prisma#3481] [Q&A] [answered]
I have something like this:
```javascript
context.prisma.$queryRaw`
  ...
  ${sql`lower(concat(P."firstName", P."middleName", P."lastName", P.suffix, T.slug, T.name, PHI."contractNumber", IP.name, IP.slug)) like '%johhan%'`}
`
```

this works but I want that part to be formed via javascript function like so:

```javascript
context.prisma.$queryRaw`
  ...
  ${searchConditions.join(" ")}
`
```

this gives me the following error:
```
Raw query failed. Code: `22P03`. Message: `db error: ERROR: incorrect binary data format in bind parameter 4`
```

If I do this:
```javascript
context.prisma.$queryRaw`
  ...
  ${sql`${searchConditions.join(" ")}`}
`
```

It stills gives me the error from above.

This is the function I'm building:
```typescript
let searchConditions: string[] = [];
args.fullSearch?.split(" ").forEach((separatedSearch, index) => {
  searchConditions.push(
    `lower(concat(P."firstName", P."middleName", P."lastName", P.suffix, T.slug, T.name, PHI."contractNumber", IP.name, IP.slug)) like '%${separatedSearch.toLowerCase()}%' ${
      index + 1 === args.fullSearch?.split(" ").length ? "" : "AND"
    }`
  );
});
```

**Comment by johhansantana** [comment, marked as answer]
Ended up doing something like this:

```typescript
const searchConditions = args.fullSearch
  ?.split(" ")
  .map((separatedSearch, index) => {
    return raw(
      ` AND lower(concat(P."firstName", P."middleName", P."lastName", P.suffix, T.slug, T.name, PHI."contractNumber", IP.name, IP.slug)) like '%${separatedSearch.toLocaleLowerCase()}%'`
    ).text;
  })
  .join(" ")
  .replace(/, AND/g, " AND");

context.prisma.$queryRaw`
  ...
  ${raw(searchConditions)}
`
```

hopefully, this is the right way to do this.

note that the `raw` is imported like this: `import { raw } from "@prisma/client";`

**Reply by Sytten** [reply]
I highly suggest you also import `join` from the client and use that so you can keep the templated string instead of reverting back to a string and having to pass the parameters as a second parameter to `queryRaw`. The documentation is very bad for that, but basically prisma uses https://github.com/blakeembrey/sql-template-tag under the hood. I was a bit worried at first that you were vulnerable to SQL injections, but I think you are fine.

**Reply by johhansantana** [reply]
SQL injections might be a problem after reading this:
https://github.com/blakeembrey/sql-template-tag#raw

But I'm not sure if using `$queryRaw` takes care of it.

**Reply by Sytten** [reply]
When you do your map, you should not add an `AND` nor convert to `.text`, but instead use the `join` helper https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-client/raw-database-access#tagged-template-helpers

**Reply by MaxiSantos** [reply]
> Ended up doing something like this:
> 
> ```ts
> const searchConditions = args.fullSearch
>   ?.split(" ")
>   .map((separatedSearch, index) => {
>     return raw(
>       ` AND lower(concat(P."firstName", P."middleName", P."lastName", P.suffix, T.slug, T.name, PHI."contractNumber", IP.name, IP.slug)) like '%${separatedSearch.toLocaleLowerCase()}%'`
>     ).text;
>   })
>   .join(" ")
>   .replace(/, AND/g, " AND");
> 
> context.prisma.$queryRaw`
>   ...
>   ${raw(searchConditions)}
> `
> ```
> 
> hopefully, this is the right way to do this.
> 
> note that the `raw` is imported like this: `import { raw } from "@prisma/client";`

Thank you. I was searching for something to let me write dynamic ´WHERE´ clause like this: https://javascript.plainenglish.io/dynamically-generating-sql-queries-using-node-js-e89d69930fcb but the use of ´Prisma.raw´ was the key.

**Reply by Sytten** [reply]
Just be careful with raw. The current solution is vulnerable to SQL injections. You need to use prepared statements with some kind of placeholder depending on your DB.

**Reply by ginger51011** [reply]
An example of using `Prisma.join` can look something like this, where `search` is the search string

```ts
const searchArray = search.toLowerCase().split(/\s+/g).map((s) => `%${s}%`);
const users = await prisma.$queryRaw`
      SELECT *
      FROM users
      WHERE lower(concat(first_name, last_name, username))
      LIKE ${
        Prisma.join(searchArray,
        ' AND lower(concat(first_name, last_name, username)) LIKE ')
      }
    `;
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
